PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/FullGfaReader.java

```diff
@@ -3,163 +3,229 @@
+   * Parse the header part of the GFA file, which contains all of the genomes.
+   *
+   * @param chars the characters of the header.
    */
-  private void handleL(Scanner sc) {
-    int parentId = sc.nextInt();
-    sc.next();
-    int childId = sc.nextInt();
-    Node parent = originalGraph.getNode(parentId);
-    Node child = originalGraph.getNode(childId);
-    parent.addOutlink(child.getId());
-    child.addInlink(parent.getId());
-    sc.nextLine();
+  private void parseHeader(char[] chars) {
+    int index = 0;
+    index = skipTillCharacter(chars, index, ':', 2) + 1;
+    int start = index;
+    while (index < chars.length) {
+      index = skipTillCharacter(chars, index, ';', 1);
+      genomes.add(new String(chars, start, index - start));
+      start = index + 1;
+      index += 2;
+    }
   }
-  
+
   /**
-   * This method extracts the single genomes from the ORI tag.
-   * @param line the full ORI tag
-   * @return an arrayList of the names of the genomes.
+   * Skips all characters till the given character is found.
+   *
+   * @param chars      the array of characters.
+   * @param startIndex the index at which to start in the array.
+   * @param ch         the character to skip to.
+   * @param amount     the amount of times to skip the character.
+   * @return the index of the first occurence of the character after the given start index.
    */
-  private ArrayList<String> extractGenomes(String line) {
-    String[] words = line.split(":");
-    if (words[0].endsWith("ORI")) {
-      String[] gens = words[2].split(";");
-      ArrayList<String> gensAr = new ArrayList<>(Arrays.asList(gens));
-      return gensAr;
+  private int skipTillCharacter(char[] chars, int startIndex, char ch, int amount) {
+    int index = startIndex;
+    for (int i = 0; i < amount; i++) {
+      ++index;
+      while (chars[index] != ch) {
+        ++index;
+      }
     }
-    return null;
+    return index;
   }
-  
+
   /**
-   * This method extracts the alignment integer from the START tag.
-   * @param line the full START tag
-   * @return the alignment integer of the tag
+   * Get a node, or create it if it doesn't exist yet.
+   *
+   * @param id the id of the node.
+   * @return the node.
    */
-  private int extractStart(String line) {
-    String[] words = line.split(":");
-    if (words[0].endsWith("START")) {
-      return Integer.parseInt(words[2]);
+  private Node getNode(int id) {
+    Node node = nodes.get(id);
+    if (node == null) {
+      node = new Node(id, 1, new ArrayList<>(), 0);
+      nodes.put(id, node);
     }
-    return 0;
+    return nodes.get(id);
   }
 
   /**
    * Method which returns the read graph.
    *
-   * @return The graph.
+   * @return the graph.
    */
   public OriginalGraph getGraph() {
-    originalGraph.setGenoms(genoms);
+    if (originalGraph == null) {
+      originalGraph = new OriginalGraph(nodes, 1, genomes);
+    }
     return originalGraph;
   }
```

Comment:
You might want to consider changing the constructor and this method. When you call the constructor now, the file will immediately be read, but not returned and you have to seperately get the result with `getGraph()`. It is more common that you initialize some reader, then call `read(...)` on it, which returns the result of reading the file. Conforming to this convention makes the code where the reader is used more obvious to others.

I see that you use a Singleton pattern here, which I don't think should be how the reader should behave (it is not in charge of storing the graph, only in reading it).

 If you really consider yours a better approach than what I suggest, please add more elaborate JavaDoc to explain the usage of the class and these key methods. For instance, tell that this method provides a Singleton object, and explain how the user should call this class in the Constructor/ Class javaDoc.